File: examples/python/minimal-order-service/adapters/inbound/rest/order_controller.py
"""
Inbound adapter: REST controller.
This is at the edge - it adapts HTTP requests to use cases.
In a real application, this would use FastAPI, Flask, etc.
"""
from dataclasses import dataclass
from application.ports.inbound.place_order_port import PlaceOrderPort, PlaceOrderCommand
from domain.order import Order


@dataclass
class PlaceOrderRequest:
    """Request DTO for placing an order."""
    amount: int
    currency: str


@dataclass
class PlaceOrderResponse:
    """Response DTO for order placement."""
    id: str
    amount: int
    currency: str
    status: str


class OrderController:
    """REST controller for order operations."""
    
    def __init__(self, place_order_port: PlaceOrderPort):
        """
        Initialize the controller.
        
        Args:
            place_order_port: Place order port
        """
        self._place_order_port = place_order_port
    
    def place_order(self, request: PlaceOrderRequest) -> PlaceOrderResponse:
        """
        Handle a place order request.
        
        Args:
            request: Place order request
        
        Returns:
            Order response
        """
        command = PlaceOrderCommand(
            amount=request.amount,
            currency=request.currency
        )
        order: Order = self._place_order_port.place_order(command)
        return PlaceOrderResponse(
            id=order.id,
            amount=order.total.amount,
            currency=order.total.currency,
            status=order.status.value
        )

File: examples/csharp/minimal-order-service/Adapters/Inbound/Rest/OrderController.cs
using OrderService.Application.Ports.Inbound;
using OrderService.Domain;

namespace OrderService.Adapters.Inbound.Rest;

/// <summary>
/// Inbound adapter: REST controller.
/// This is at the edge - it adapts HTTP requests to use cases.
/// In a real application, this would use ASP.NET Core, Minimal API, etc.
/// </summary>
public class OrderController
{
    private readonly IPlaceOrderPort _placeOrderPort;

    public OrderController(IPlaceOrderPort placeOrderPort)
    {
        _placeOrderPort = placeOrderPort;
    }

    public PlaceOrderResponse PlaceOrder(PlaceOrderRequest request)
    {
        var command = new PlaceOrderCommand(request.Amount, request.Currency);
        var order = _placeOrderPort.PlaceOrder(command);
        
        return new PlaceOrderResponse(
            order.Id,
            order.Total.Amount,
            order.Total.Currency,
            order.Status.ToString()
        );
    }
}

// DTOs for REST layer
public record PlaceOrderRequest(long Amount, string Currency);

public record PlaceOrderResponse(
    string Id,
    long Amount,
    string Currency,
    string Status
);
